File: ruby/two-bucket/two_bucket.rb
class TwoBucket
  attr_reader :moves, :goal_bucket, :other_bucket

  def initialize(bucket_one_volume, bucket_two_volume, goal, start_bucket)
    @bucket_volumes = { one: bucket_one_volume, two: bucket_two_volume }
    @goal = goal
    @start_bucket = start_bucket
    @goal_state = calculate_moves
    @moves ||= @goal_state[2]
    @goal_bucket = @goal_state[0] == @goal ? "one" : "two"
    @other_bucket = @goal_state[0] == @goal ? @goal_state[1] : @goal_state[0]
  end

  def calculate_moves
    initial_state =
      (
        if @start_bucket == "one"
          [@bucket_volumes[:one], 0, 1]
        else
          [0, @bucket_volumes[:two], 1]
        end
      )
    queue = [initial_state]
    visited_states = Set.new

    until queue.empty?
      current_state = queue.shift

      return current_state if current_state[0] == @goal || current_state[1] == @goal

      visited_states.add(current_state[..1])
      next_states = next_states(current_state)

      next_states.each do |state|
        queue.push(state) unless visited_states.include?(state[..1]) || !valid_state?(state)
      end
    end
  end

  def next_states(current_state)
    bucket_one, bucket_two, moves = current_state
    next_states = []

    # Fill bucket one
    next_states.push([@bucket_volumes[:one], bucket_two, moves + 1])

    # Fill bucket two
    next_states.push([bucket_one, @bucket_volumes[:two], moves + 1])

    # Empty bucket one
    next_states.push([0, bucket_two, moves + 1])

    # Empty bucket two
    next_states.push([bucket_one, 0, moves + 1])

    # Pour from bucket one to bucket two
    pour = [bucket_one, bucket_two, moves + 1]
    pour[0] -= [@bucket_volumes[:two] - bucket_two, bucket_one].min
    pour[1] += [@bucket_volumes[:two] - bucket_two, bucket_one].min
    next_states.push(pour)

    # Pour from bucket two to bucket one
    pour = [bucket_one, bucket_two, moves + 1]
    pour[1] -= [@bucket_volumes[:one] - bucket_one, bucket_two].min
    pour[0] += [@bucket_volumes[:one] - bucket_one, bucket_two].min
    next_states.push(pour)

    next_states
  end

  def valid_state?(state)
    if @start_bucket == "one"
      state[0].positive? || state[1] < @bucket_volumes[:two]
    else
      state[0] < @bucket_volumes[:one] || state[1].positive?
    end
  end
end

File: typescript/two-bucket/two-bucket.ts
type Buckets = { one: number; two: number };

type BucketsState = {
  buckets: Buckets;
  moves: number;
} | null;

export class TwoBucket {
  goalState?: BucketsState;
  buckOne: number;
  buckTwo: number;
  starterBuck: string;
  _goalBucket?: string;
  _otherBucket?: number;

  constructor(
    buckOne: number,
    buckTwo: number,
    goal: number,
    starterBuck: string,
  ) {
    this.buckOne = buckOne;
    this.buckTwo = buckTwo;
    this.starterBuck = starterBuck;
    this.goalState = this.findGoalState(buckOne, buckTwo, goal, starterBuck);
    this._goalBucket = this.goalState?.buckets.one === goal ? "one" : "two";
    this._otherBucket =
      this._goalBucket === "one"
        ? this.goalState?.buckets.two
        : this.goalState?.buckets.one;
  }

  moves(): number {
    if (!this.goalState) {
      throw new Error("Goal not reachable");
    }

    return this.goalState.moves;
  }

  get goalBucket(): string {
    return this._goalBucket || "";
  }

  get otherBucket(): number {
    return this._otherBucket || 0;
  }

  private findGoalState(
    buckOne: number,
    buckTwo: number,
    goal: number,
    starterBuck: string,
  ): BucketsState | null {
    const states: BucketsState[] = [];
    const visited: Set<string> = new Set();

    states.push(this.findInitialState(buckOne, buckTwo, starterBuck));

    while (states.length) {
      let state = states.shift();
      let nextStates;

      if (state?.buckets.one === goal || state?.buckets.two === goal) {
        return state;
      }

      if (state) {
        visited.add(this.stringifyBuckets(state.buckets));
        nextStates = this.getNextStates(state);

        for (let next of nextStates) {
          if (next && !visited.has(this.stringifyBuckets(next.buckets))) {
            states.push(next);
          }
        }
      }
    }

    return null;
  }

  private findInitialState(
    buckOne: number,
    buckTwo: number,
    starterBuck: string,
  ): BucketsState {
    if (starterBuck === "one") {
      return { buckets: { one: buckOne, two: 0 }, moves: 1 };
    }
    return { buckets: { one: 0, two: buckTwo }, moves: 1 };
  }

  private getNextStates(state: BucketsState): BucketsState[] {
    let states = [
      this.fillOne(state),
      this.fillTwo(state),
      this.emptyOne(state),
      this.emptyTwo(state),
      this.pourOneIntoTwo(state),
      this.pourTwoIntoOne(state),
    ];

    return this.removeInvalidStates(states);
  }

  private fillOne(state: BucketsState): BucketsState {
    if (state?.buckets.one === this.buckOne) {
      return state;
    }

    return {
      buckets: { one: this.buckOne, two: Number(state?.buckets.two) },
      moves: Number(state?.moves) + 1,
    };
  }

  private fillTwo(state: BucketsState): BucketsState {
    if (state?.buckets.two === this.buckTwo) {
      return state;
    }

    return {
      buckets: { one: Number(state?.buckets.one), two: this.buckTwo },
      moves: Number(state?.moves) + 1,
    };
  }

  private emptyOne(state: BucketsState): BucketsState {
    if (state?.buckets.one === 0) {
      return state;
    }

    return {
      buckets: { one: 0, two: Number(state?.buckets.two) },
      moves: Number(state?.moves) + 1,
    };
  }

  private emptyTwo(state: BucketsState): BucketsState {
    if (state?.buckets.two === 0) {
      return state;
    }

    return {
      buckets: { one: Number(state?.buckets.one), two: 0 },
      moves: Number(state?.moves) + 1,
    };
  }

  private pourOneIntoTwo(state: BucketsState): BucketsState {
    let remainingVolume = this.buckTwo - Number(state?.buckets.two);
    let volumeToPour = Math.min(Number(state?.buckets.one), remainingVolume);

    if (volumeToPour === 0) {
      return state;
    }

    return {
      buckets: {
        one: Number(state?.buckets.one) - volumeToPour,
        two: Number(state?.buckets.two) + volumeToPour,
      },
      moves: Number(state?.moves) + 1,
    };
  }

  private pourTwoIntoOne(state: BucketsState): BucketsState {
    let remainingVolume = this.buckOne - Number(state?.buckets.one);
    let volumeToPour = Math.min(Number(state?.buckets.two), remainingVolume);

    if (volumeToPour === 0) {
      return state;
    }

    return {
      buckets: {
        one: Number(state?.buckets.one) + volumeToPour,
        two: Number(state?.buckets.two) - volumeToPour,
      },
      moves: Number(state?.moves) + 1,
    };
  }

  private stringifyBuckets(buckets: Buckets): string {
    return `${buckets.one},${buckets.two}`;
  }

  private removeInvalidStates(states: BucketsState[]): BucketsState[] {
    return states.filter((state: BucketsState) => {
      if (this.starterBuck === "one") {
        return state?.buckets.one !== 0 || state?.buckets.two !== this.buckTwo;
      } else {
        return state?.buckets.one !== this.buckOne || state?.buckets.two !== 0;
      }
    });
  }
}
